lumeer-storage/lumeer-storage-mongodb/src/main/java/io/lumeer/storage/mongodb/dao/system/MongoOrganizationDao.java

```diff
@@ -53,4 +69,73 @@ public Organization getOrganizationByCode(final String organizationCode) {
       }
       return organization;
    }
+
+   private Query<MongoOrganization> createOrganizationQuery(DatabaseQuery query) {
+      Query<MongoOrganization> organizationQuery = datastore.createQuery(MongoOrganization.class);
+
+      List<Criteria> criteria = new ArrayList<>();
+      criteria.add(createUserCriteria(organizationQuery, query.getUser()));
+      query.getGroups().forEach(group -> criteria.add(createGroupCriteria(organizationQuery, group)));
+      organizationQuery.or(criteria.toArray(new Criteria[] {}));
+
+      return organizationQuery;
+   }
+
+   private Criteria createUserCriteria(Query<MongoOrganization> organizationQuery, String user) {
+      return organizationQuery.criteria(MongoOrganization.PERMISSIONS + "." + MongoPermissions.USER_ROLES)
+                              .elemMatch(createPermissionQuery(user));
+   }
+
+   private Query<MongoPermission> createPermissionQuery(String name) {
+      return datastore.createQuery(MongoPermission.class)
+                      .filter(MongoPermission.NAME, name)
+                      .field(MongoPermission.ROLES).in(Collections.singleton(Role.READ.toString()));
+   }
+
+   private Criteria createGroupCriteria(Query<MongoOrganization> organizationQuery, String group) {
+      return organizationQuery.criteria(MongoOrganization.PERMISSIONS + "." + MongoPermissions.GROUP_ROLES)
+                              .elemMatch(createPermissionQuery(group));
+   }
+
+   private FindOptions createFindOptions(DatabaseQuery query) {
+      FindOptions findOptions = new FindOptions();
+      Integer page = query.getPage();
+      Integer pageSize = query.getPageSize();
+
+      if (page != null && pageSize != null) {
+         findOptions.skip(page * pageSize)
+                    .limit(pageSize);
+      }
+
+      return findOptions;
+   }
+
+   @Override
+   public List<Organization> getOrganizations(final DatabaseQuery query) {
+      Query<MongoOrganization> organizationQuery = createOrganizationQuery(query);
+      FindOptions findOptions = createFindOptions(query);
+
+      return new ArrayList<>(organizationQuery.asList(findOptions));
+   }
+
+   @Override
+   public void deleteOrganization(final String organizationCode) {
+      WriteResult writeResult = datastore.delete(
+            datastore.createQuery(MongoOrganization.class)
+                     .field(MongoOrganization.CODE).equal(organizationCode));
+      if (writeResult.getN() != 1) {
+         throw new WriteFailedException(writeResult);
+      }
+   }
+
+   @Override
+   public Organization editOrganization(final String organizationCode, final Organization organization) {
+      MongoOrganization mongoOrganization = new MongoOrganization(organization);
+      String id = datastore.createQuery(MongoOrganization.class)
+               .field(MongoOrganization.CODE).equal(organizationCode).get().getId();
```

Comment:
Maybe you should use `getOrganizationByCode` here so there is no duplicated code and the exception is thrown when the organization is not found.